File: RealtimeDatabase/Models/Prefilter/WherePrefilter.cs
using RealtimeDatabase.Internal;
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;
using RealtimeDatabase.Helper;

// ReSharper disable PossibleMultipleEnumeration

namespace RealtimeDatabase.Models.Prefilter
{
    public class WherePrefilter : IPrefilter
    {
        public string CompareFunctionString { get; set; }

        public object[] ContextData { get; set; }

        private Func<object, bool> predicate;

        public IEnumerable<object> Execute(IEnumerable<object> array)
        {
            if (array.Any())
            {
                if (predicate == null)
                {
                    predicate = CompareFunctionString.CreateBoolFunction(ContextData);
                }

                return array.Where(predicate);
            }

            return array;
        }
    }
}

File: DemoClient/projects/ng-realtime-database/src/lib/models/prefilter/where-prefilter.ts
import {IPrefilter} from './iprefilter';

export class WherePrefilter<T> implements IPrefilter<T> {
  prefilterType = 'WherePrefilter';
  compareFunction: (x: T) => boolean;
  compareFunctionString: string;
  contextData: { [key: string]: string };

  constructor(compareFunction: (x: T) => boolean, contextData?: { [key: string]: any }) {
    this.compareFunction = compareFunction;
    this.compareFunctionString = compareFunction.toString();

    if (contextData) {
      this.contextData = {};

      for (const key of Object.keys(contextData)) {
        this.contextData[key] = JSON.stringify(contextData[key]);
      }
    }
  }

  public execute(values: T[]) {
    return values;
  }

  public hash() {
    return `${this.prefilterType},${this.compareFunctionString},${JSON.stringify(this.contextData)}`;
  }
}
